python/cudf/cudf/io/orc.py

```diff
@@ -294,6 +294,26 @@ def read_orc_stripe(orc_file, stripe, columns):
 def to_orc(df, fname, compression=None, enable_statistics=True, **kwargs):
     """{docstring}"""
 
+    if any(
+        isinstance(col, cudf.core.column.ListColumn)
+        for col in df._data.columns
+    ):
+        raise NotImplementedError(
+            "Writing to ORC format is not yet supported with list columns."
+        )
+
+    if any(
+        isinstance(col, cudf.core.column.CategoricalColumn)
+        for col in df._data.columns
+    ) or isinstance(df.index, cudf.CategoricalIndex):
+        df = df.copy(deep=False)
+        for col_name, col in df._data.items():
+            if isinstance(col, cudf.core.column.CategoricalColumn):
+                df._data[col_name] = col.astype(col.cat().categories.dtype)
+
+        if isinstance(df.index, cudf.CategoricalIndex):
+            df.index = df.index.astype(df.index.categories.dtype)
```

Comment:
Does the ORC spec support dictionary encoded columns? I'm wondering if this is unexpected behavior for the user as these columns wouldn't round trip properly.

Reply:
From the spec it appears there is dictionary support but I would fallback to @vuule or @rgsl888prabhu on this.

The reason I'm converting categorical dtype to underlying dtype is currently in our orc writer we are writing categorical codes as opposed to the true values - This is a similar stopgap used incase of csv writer where we were having similar issue.

Reply:
ORC spec supports dictionary type columns but they use name [`MAP Columns`](https://orc.apache.org/specification/ORCv1/), but currently cudf ORC writer doesn't support writing that type.

Reply:
Isn't Map more like key-value pairs as opposed to dictionary encoded columns? It looks like the ORC spec supports dictionary encoding where maybe it makes sense for us to read dictionary encoded ORC columns back as CategoricalColumns in the future?

Regardless, I don't think we want to decode the CategoricalColumn here.

Reply:
ahh, yes. MAP is different compared to Categorical/Dictionary Column, my bad.

I don't see any dictionary support in ORC spec even though it has dictionary encoding.
```
enum Kind {
 BOOLEAN = 0;
 BYTE = 1;
 SHORT = 2;
 INT = 3;
 LONG = 4;
 FLOAT = 5;
 DOUBLE = 6;
 STRING = 7;
 BINARY = 8;
 TIMESTAMP = 9;
 LIST = 10;
 MAP = 11;
 STRUCT = 12;
 UNION = 13;
 DECIMAL = 14;
 DATE = 15;
 VARCHAR = 16;
 CHAR = 17;
 }
```

Reply:
So should we throw for now if someone paases a dictionary column at python layer?

Reply:
What's the current behavior?

Reply:
Currently we are writing the codes column

Reply:
Alright well that's very wrong and no one has reported it, so I think we should make the behavior to throw instead of silently writing incorrect data. Once cudf python categorical columns are hooked up to libcudf dictionary columns we can explore further.

Reply:
Sure, Updated the code to raise error for categorical data.